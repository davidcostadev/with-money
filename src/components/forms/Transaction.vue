<template>
  <form>
    <md-card>
      <md-card-content>
        <md-field>
          <label for="name">Name</label>
          <md-input v-model="nameMutable" id="name" />
        </md-field>
        <md-field>
          <input type="text" id="value" v-model="valueMutable" class="md-input" />
          <md-icon>attach_money</md-icon>
        </md-field>
        <md-datepicker v-model="transactionDateMutable" id="transactionDate">
          <label>Transaction Day</label>
        </md-datepicker>
        <md-field>
          <label for="type">Type</label>
          <md-select v-model="typeMutable" name="type" id="type">
            <md-option value="in" id="type-in">In</md-option>
            <md-option value="out" id="type-out">Out</md-option>
          </md-select>
        </md-field>
        <md-field>
          <label for="AccountId">Account</label>
          <md-select v-model="AccountIdMutable" name="AccountId" id="account">
            <md-option
            v-for="account in accounts"
            :key="account.id"
            :value="account.id"
            :id="`account-${account.id}`"
            >
              {{account.name}}
            </md-option>
          </md-select>
        </md-field>
        <md-field>
          <label for="CategoryId">Category</label>
          <md-select v-model="CategoryIdMutable" name="CategoryId" id="category">
            <md-option
            v-for="category in categories"
            :key="category.id"
            :value="category.id"
            :id="`category-${category.id}`"
            >
              {{category.name}}
            </md-option>
          </md-select>
        </md-field>
        <div>
          <md-switch v-model="isPaidMutable" class="md-primary" id="isPaid">IsPaid?</md-switch>
        </div>
      </md-card-content>
      <md-card-actions>
        <md-button
        class="md-primary"
        @click="save"
        :disabled="isLoading"
        id="save"
      >Save</md-button>
      </md-card-actions>
    </md-card>
  </form>
</template>

<script>
import Accounts from '../../services/accounts';
import Categories from '../../services/categories';

export default {
  props: {
    name: {
      type: String,
      default: '',
    },
    value: {
      type: String,
      default: '',
    },
    AccountId: {
      type: Number,
      default: 0,
    },
    CategoryId: {
      type: Number,
      default: 0,
    },
    type: {
      type: String,
      default: 'out',
    },
    isPaid: {
      type: Boolean,
      default: false,
    },
    transactionDate: {
      type: [String, Date],
      default: () => new Date(),
    },
    onSave: {
      type: Function,
      required: true,
    },
  },
  data() {
    return {
      accounts: [],
      categories: [],
      isLoading: false,
      nameMutable: this.name,
      valueMutable: this.value,
      transactionDateMutable: this.transactionDate,
      typeMutable: this.type,
      isPaidMutable: this.isPaid,
      AccountIdMutable: this.AccountId,
      CategoryIdMutable: this.CategoryId,
    };
  },
  methods: {
    async save() {
      this.isLoading = true;

      await this.onSave({
        name: this.nameMutable,
        value: this.valueMutable,
        AccountId: this.AccountIdMutable,
        CategoryId: this.CategoryIdMutable,
        type: this.typeMutable,
        isPaid: this.isPaidMutable,
        transactionDate: this.transactionDateMutable,
      });

      this.isLoading = false;
    },

    async getAccounts() {
      const { data } = await Accounts.list();
      this.accounts = data;
    },
    async getCategories() {
      const { data } = await Categories.list({
        type: this.typeMutable,
      });

      this.categories = data;
    },
  },
  async created() {
    await this.getAccounts();
    await this.getCategories();
  },
};
</script>
