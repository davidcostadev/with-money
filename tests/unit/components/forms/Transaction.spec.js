import { createLocalVue, shallowMount, mount } from '@vue/test-utils';
import VueMaterial from 'vue-material';
import Transaction from '../../../../src/components/forms/Transaction';

const localVue = createLocalVue();

localVue.use(VueMaterial);

const mockAccounts = {
  data: {
    data: [
      {
        id: 2,
        UserId: 1,
        name: 'Carteira',
        type: 'wallet',
        initalValue: '650.00',
        createdAt: '2018-08-04T14:33:21.000Z',
        updatedAt: '2018-08-04T14:33:21.000Z',
      },
      {
        id: 1,
        UserId: 1,
        name: 'Banco Inter',
        type: 'checking_account',
        initalValue: '505.18',
        createdAt: '2018-08-04T14:33:21.000Z',
        updatedAt: '2018-08-04T14:33:21.000Z',
      },
    ],
    pagination: {
      totalItems: 2,
      currentPage: 1,
      perPage: 100,
      totalPages: 1,
      nextPage: null,
      previousPage: null,
    },
  },
};

const mockCategories = {
  data: {
    data: [
      {
        id: 2,
        UserId: 1,
        name: 'Lanche',
        type: 'out',
        initalValue: '650.00',
        createdAt: '2018-08-04T14:33:21.000Z',
        updatedAt: '2018-08-04T14:33:21.000Z',
      },
      {
        id: 1,
        UserId: 1,
        name: 'Salario',
        type: 'in',
        createdAt: '2018-08-04T14:33:21.000Z',
        updatedAt: '2018-08-04T14:33:21.000Z',
      },
    ],
    pagination: {
      totalItems: 2,
      currentPage: 1,
      perPage: 100,
      totalPages: 1,
      nextPage: null,
      previousPage: null,
    },
  },
};

jest.mock('../../../../src/services/accounts', () => ({
  list: () => Promise.resolve(mockAccounts),
}));

jest.mock('../../../../src/services/categories', () => ({
  list: () => Promise.resolve(mockCategories),
}));

describe('Transaction Component', () => {
  describe('Render', () => {
    it('default', () => {
      const onSave = jest.fn();


      const wrapper = shallowMount(Transaction, {
        localVue,
        propsData: {
          onSave,
          transactionDate: '2018-08-04T14:33:21.000Z',
        },
      });

      expect(wrapper.isVueInstance).toBeTruthy();
      expect(wrapper.element).toMatchSnapshot();
    });
  });

  describe('Behavior', () => {
    it('onSave', () => {
      const onSave = jest.fn();
      const wrapper = mount(Transaction, {
        localVue,
        propsData: {
          onSave,
        },
        sync: false,
      });

      wrapper.find('#save').trigger('click');

      expect(onSave.mock.calls[0][0]).toHaveProperty('name');
      expect(onSave.mock.calls[0][0]).toHaveProperty('value');
      expect(onSave.mock.calls[0][0]).toHaveProperty('AccountId');
      expect(onSave.mock.calls[0][0]).toHaveProperty('isPaid');
      expect(onSave.mock.calls[0][0]).toHaveProperty('type');
      expect(onSave.mock.calls[0][0]).toHaveProperty('transactionDate');

      expect(onSave.mock.calls[0][0].name).toBe('');
      expect(onSave.mock.calls[0][0].value).toBe('');
      expect(onSave.mock.calls[0][0].AccountId).toBe(0);
      expect(onSave.mock.calls[0][0].isPaid).toBe(false);
      expect(onSave.mock.calls[0][0].type).toBe('out');
      expect(onSave.mock.calls[0][0].transactionDate instanceof Date).toBe(true);
    });
  });
});
