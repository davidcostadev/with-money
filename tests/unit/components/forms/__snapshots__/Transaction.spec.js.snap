// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transaction Component Render default 1`] = `
<form>
  <mdcard-stub>
    <mdcardcontent-stub>
      <mdfield-stub
        mdcounter="true"
        mdtogglepassword="true"
      >
        <label
          for="name"
        >
          Name
        </label>
         
        <mdinput-stub
          id="name"
          type="text"
        />
      </mdfield-stub>
       
      <mdfield-stub
        mdcounter="true"
        mdtogglepassword="true"
      >
        <input
          class="md-input"
          id="value"
          type="text"
        />
         
        <mdicon-stub>
          attach_money
        </mdicon-stub>
      </mdfield-stub>
       
      <mddatepicker-stub
        id="transactionDate"
        mddebounce="1000"
        mdopenonfocus="true"
        mdoverridenative="true"
        value="Sat Aug 04 2018 11:33:21 GMT-0300 (-03)"
      >
        <label>
          Transaction Day
        </label>
      </mddatepicker-stub>
       
      <mdfield-stub
        mdcounter="true"
        mdtogglepassword="true"
      >
        <label
          for="type"
        >
          Type
        </label>
         
        <mdselect-stub
          id="type"
          name="type"
        >
          <mdoption-stub
            id="type-in"
            value="in"
          >
            In
          </mdoption-stub>
           
          <mdoption-stub
            id="type-out"
            value="out"
          >
            Out
          </mdoption-stub>
        </mdselect-stub>
      </mdfield-stub>
       
      <mdfield-stub
        mdcounter="true"
        mdtogglepassword="true"
      >
        <label
          for="AccountId"
        >
          Account
        </label>
         
        <mdselect-stub
          id="account"
          name="AccountId"
        />
      </mdfield-stub>
       
      <mdfield-stub
        mdcounter="true"
        mdtogglepassword="true"
      >
        <label
          for="CategoryId"
        >
          Category
        </label>
         
        <mdselect-stub
          id="category"
          name="CategoryId"
        />
      </mdfield-stub>
       
      <div>
        <mdswitch-stub
          class="md-primary"
          id="isPaid"
        >
          IsPaid?
        </mdswitch-stub>
      </div>
    </mdcardcontent-stub>
     
    <mdcardactions-stub
      mdalignment="right"
    >
      <mdbutton-stub
        class="md-primary"
        id="save"
        type="button"
      >
        Save
      </mdbutton-stub>
    </mdcardactions-stub>
  </mdcard-stub>
</form>
`;
